# Title     : Graphs Library
# Objective : Genaric Functions used for graphs
# Created by: alexm
# Created on: 11/8/2020

# Given a datatable generated by assemble-with-stats a set of values for the x-axis
# and a value to label the x-axis with generates a plot representing the data
plot.length.n50 <- function(data, x, xlab) {
  plot(x, data$total.length, col = "black", type = "l",
       ylim = c(0, max(data$total.length)), xlab = xlab, ylab = "Length (bp)")
  lines(x, data$N50, col = "blue")

  title(main = paste("Genome Quality Vs", xlab))
  legend("bottomleft",
         legend = c("Total Genome Length", "N50"),
         col = c("black", "blue"),
         pch = 19,
         inset = c(0.1, 0.1))
}

# Plots the number of contigs generated against the provided x-axis
plot.num.contigs <- function(data, x, xlab) {
  plot(x, data$X..contigs, type = "b",
       ylim = c(0, max(data$X..contigs)), ylab = "# Contigs", xlab = xlab)
  title(main = paste("Number of Contigs Vs", xlab))
}

# Plots all the run times against the provided x-axis, also take a position for the legend
plot.runtimes <- function(data, x, xlab, lpos, linset) {
  ty <- "o"
  cols <- c("black", "blue", "cyan", "purple", "orange", "red")

  start <- as.POSIXct("00:00:00", format = "%H:%M:%S")
  real.time <- as.POSIXct(data$Real.Flye.Assembly.time, format = "%M:%S")
  seconds <- function(r) { as.POSIXct(r, format = "%s", origin = start) }

  plot(x, real.time, ylim = c(start, max(real.time)), col = cols[1], type = ty,
       ylab = "Time", xlab = xlab)
  lines(x, seconds(data$Quorum.Error.Correcting.time), col = cols[6], type = ty)
  lines(x, seconds(data$Estimating.Genome.Size.time), col = cols[5], type = ty)
  lines(x, seconds(data$Super.Reads.time), col = cols[4], type = ty)
  lines(x, seconds(data$Mega.Reads.time), col = cols[3], type = ty)
  lines(x, seconds(data$Consensus.Gap.Joining.time), col = cols[2], type = ty)

  title(main = paste("Runtimes Vs", xlab))
  legend(lpos, inset = linset, pch = 1,
         col = cols,
         legend = c("Real/Flye Assembly",
                    "Consensus Gap Joining",
                    "Mega-Reads",
                    "Super-Reads",
                    "Estimating Genome Size",
                    "Quorum Error Correcting"))
}